docs/design-documents/features/storage/TDBStore/TDBStore_design.md

```diff
@@ -0,0 +1,497 @@
+
+Functionality, as defined by KVStore, includes the following:
+- Initialization & reset
+- Core actions: get, set & remove
+- Incremental set actions
+- Iterator actions
+
+### Class header
+
+TDBStore has the following header:
+
+```C++
+class TDBStore : KVStore {
+
+public:
+    TDBSTore(size_t max_keys, BlockDevice *bd = 0);
+    virtual ~TDBSTore();
+    	 
+    // Initialization and reset
+    virtual int init();
+    virtual int deinit();
+    virtual int reset();
+
+    // Core API
+    virtual int set(const char *key, const void *buffer, size_t size, uint32_t create_flags);
+    virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size = NULL, size_t offset = 0);
+    virtual int get_info(const char *key, info_t *info);
+    virtual int remove(const char *key);
+ 
+    // Incremental set API
+    virtual int set_start(set_handle_t *handle, const char *key, size_t final_data_size, uint32_t create_flags);
+    virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);
+    virtual int set_finalize(set_handle_t handle);
+ 
+    // Key iterator
+    virtual int iterator_open(iterator_t *it, const char *prefix = NULL);
+    virtual int iterator_next(iterator_t it, char *key, size_t key_size);
+    virtual int iterator_close(iterator_t it);
+    
+    // Reserved space APIs
+    virtual int reserved_space_set(void *data);
+    virtual int reserved_space_get(void *data);
+    
+private:
+    Mutex _mutex;
+    void *_ram_table;
+    size_t *_max_keys;
+    size_t *_num_keys;
+    BlockDevice *_bd;
+    bd_addr_t _free_space_offset;
+    BufferedBlockDevice *_buff_bd;
+    bool _is_initialized;
+    int _active_area;
+    
+    // Important internal functions
+    
+    // find record offset in flash
+    int find_record(const char *key, uint32_t *hash, bd_size_t *bd_offset, size_t ram_table_ind);
+    
+    // garbage collection
+    int garbage_collection(const char *key, const void *buffer, size_t size, uint32_t create_flags);
+}
+```
+
+### Important data structures
+
+```C++
+// RAM table entry
+typedef struct {
+    uint32_t  hash;
+    bd_size_t bd_offset;
+} ram_table_entry_t;
+
+// Record header
+typedef struct {
+    uint32_t magic;
+    uint16_t header_size;
+    uint16_t revision;
+    uint32_t user_flags;
+    uint16_t int_flags;
+    uint16_t key_size;
+    uint32_t data_size;
+    uint32_t crc;
+} record_header_t;
+
+// incremental set handle
+typedef struct {
+    bd_size_t bd_base_offset;
+    bd_size_t bd_curr_offset;
+    size_t key_size;
+    size_t data_size;
+    uint32_t create_flags;
+    uint32_t crc;
+} inc_set_handle_t;
+
+// iterator handle
+typedef struct {
+    size_t ram_table_ind;
+    char *prefix;
+} key_iterator_handle_t;
+```
+
+
+### Initialization and reset
+
+**init function**
+
+Header:  
+`virtual int init();`
+
+Pseudo code:  
+- if `_is_initialized` return OK
+- Take `_mutex` 
+- Allocate `_ram_table` as an array of `_num_keys`  
```

Comment:
It may be worth noting up in the RAM table section that the RAM table is fixed size at initialization time.

That being said, couldn't you make this RAM table dynamic with amortized doubling? This can be done pretty cheaply:
``` cpp
// constructor, optional initial table size, defaults to 16 keys
TDBStore::TDBStore(int initial_table_capacity = 16) {
    _ram_table = new key_struct[initial_table_capacity];
    _ram_table_capacity = initial_table_capacity;
    _ram_table_size = 0;
}

// ram-table insert function
void TDBStore::ram_table_insert(int pos, key_struct key) {
    if (_ram_table_size == _ram_table_capacity) {
        // double the ram_table
        int new_table_capacity = 2*_ram_table_capacity;
        key_struct *new_table = new key_struct[new_table_capacity];
        // copy over table contents
        memcpy(new_table, _ram_table, _ram_table_capacity * sizeof(key_struct));
        memset(new_table + _ram_table_capacity, 0, _ram_table_capacity);
        // free and replace old table
        delete[] _ram_table;
        _ram_table = new_table;
        _ram_table_capacity = new_table_capacity;
    }

    // insert new key
    memmove(&_ram_table[pos+1], &_ram_table[pos],
        (_ram_table_size-pos) * sizeof(key_struct));
    _ram_table[pos] = key;
    _ram_table_size += 1;
}

// destructor deletes keys, no other logic needed for removes (most tables don't decrease in size)
TDBStore::~TDBStore() {
    delete[] _ram_table;
}
```

Reply:
Yes it is an option. Only down size is the peak memory allocation. If we have 64 entries (8 bytes each), then adding a new key would result in a peak allocation of 1KB for this table, comparing to half as much if no doubling is implemented. But otherwise it's the better choice of course.

Reply:
Note: This is just a suggestion, not a blocker. Can be added in 5.12.

If you let the user pass the capacity in the constructor, then you can avoid over-allocating if the user knows how many keys they will have. So I think the only downside is code size.